### PR TITLE
Add prefix or suffix sampler numbering for HTTP Test Script Recorder

### PR DESCRIPTION
**Pull request by vdaburon**
## Description
Change GUI for HTTP(s)Test Script Recorder add choice list for sampler numbering in prefix or suffix

Add choice list in with
    Prefix number (start) (Default)
    Suffix number (at the end)
    Without numbering

Add text field to set the next number start at 1

Add text field to set the number format (String format) String format syntax default %03d

## Motivation and Context
I preferred the nummbering of the samplers like the version of JMeter 3.3 and there was no choice of selection of the suffix or prefix number

Quite often I had to leave JMeter to have a sampler numbering that starts at 1.
Now you can set the next number.

## How Has This Been Tested?
Select all different possibilities for naming samplers (Prefix name, Transaction name, prefix numbering, suffix numbering, no numbering, change next value, change numbering format)

## Screenshots (if appropriate):
example_record_numbering

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist:

- [X ] My code follows the [code style][style-guide] of this project.
- [ ] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

**Comment by vdaburon** [comment]
Could this pull request will be merge ?

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/gui/ProxyControlGui.java [hunk 1]
```diff
@@ -412,10 +445,25 @@ private void populateTable(PowerTableModel pModel, PropertyIterator iter) {
...
             JComboBox combo = (JComboBox) e.getSource();
-            if(HTTP_SAMPLER_NAMING_MODE.equals(combo.getName())){
+            if(HTTP_SAMPLER_NAMING_MODE.equals(combo.getName())) {
                 model.setHTTPSampleNamingMode(httpSampleNamingMode.getSelectedIndex());
+            }
+
+            if(HTTP_SAMPLER_NUMBERING_MODE.equals(combo.getName())) {
+                model.setHttpSampleNumberingMode(httpSampleNumberingMode.getSelectedIndex());
+                if (httpSampleNumberingMode.getSelectedIndex() == 0) {
+                	JMeterUtils.setProperty("proxy.number.mode", "prefix");
```
Comment (pmouawad) [inline src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/gui/ProxyControlGui.java:455]: Why this and not using setter/getters ?

**Review by pmouawad — CHANGES_REQUESTED** [review]
Hello,
Thanks for PR.
See my remarks which require changes before considering merge.
Thanks

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/gui/ProxyControlGui.java [hunk 2]
```diff
@@ -442,6 +483,12 @@ public void actionPerformed(ActionEvent action) {
             }
         } else if (command.equals(ACTION_RESTART)) {
             model.stopProxy();
+            // what is the last number for numbering samplers, ex : 96
+            int iRequestNumber = model.getRequestNumberFromSamplerCreator();
```
Comment (FSchumacher) [inline src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/gui/ProxyControlGui.java:487]: If the *`i`* at the beginning of `iRequestNumber` is used to indicate an integer, I would remove it and name the variable `requestNumber`. That is the way variables are named in JMeter most of the time. Most developers around JMeter will probably use an IDE, which shows the type of the variable anyway.

File: xdocs/usermanual/component_reference.xml [hunk 3]
```diff
@@ -6720,13 +6719,34 @@ Both Chrome and Internet Explorer use the same trust store for certificates.
         By default it also removes <code>If-Modified-Since</code> and <code>If-None-Match</code> headers.
         These are used to determine if the browser cache items are up to date;
         when recording one normally wants to download all the content.
-        To change which additional headers are removed, define the JMeter property <code>proxy.headers.remove</code>
+        To change which additional headers are removed, define the JMeter property <code>proxy.heade'rs.remove</code>
```
Comment (FSchumacher) [inline xdocs/usermanual/component_reference.xml:6722]: Seems to be a strange name for a property. Is `proxy.heade'rs` really correct?

File: xdocs/usermanual/component_reference.xml [hunk 4]
```diff
@@ -6720,13 +6719,34 @@ Both Chrome and Internet Explorer use the same trust store for certificates.
...
         as a comma-separated list of headers.
         </property>
         <property name="Add Assertions" required="Yes">Add a blank assertion to each sampler?</property>
         <property name="Regex Matching" required="Yes">Use Regex Matching when replacing variables? If checked replacement will use word boundaries, i.e. it will only replace word matching values of variable, not part of a word. A word boundary follows Perl5 definition and is equivalent to <code>\b</code>. More information below in the paragraph about "<code>User Defined Variable replacement</code>".</property>
         <property name="Prefix/Transaction name" required="No">Add a prefix to sampler name during recording (Prefix mode). Or replace sampler name by user chosen name (Transaction name)</property>
-        <property name="Create new transaction after request (ms)">Inactivity time between two requests needed to consider them in two separate groups.</property>
+		<property name="Numbering Sampler Choice" required="Yes">Select the numbering mode for sampler name<br/>
```
Comment (FSchumacher) [inline xdocs/usermanual/component_reference.xml:6728]: Looks like a mismatch in spaces at the front. I think the `property` tags should match up.

File: xdocs/usermanual/component_reference.xml [hunk 5]
```diff
@@ -6720,13 +6719,34 @@ Both Chrome and Internet Explorer use the same trust store for certificates.
...
+		<ul>
+             <li><code>Without numbering</code> no numbering, ex : '/a.png'</li>
```
Comment (FSchumacher) [inline xdocs/usermanual/component_reference.xml:6730]: *ex : ...* is probably a French abbreviation (it means *excercise* in English). In English you can use *e.g.*, or *for example*. Also note, that there is no space before a colon in English typography. And regarding the markup of `/a.png`, I would use a `code` block.

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java [hunk 6]
```diff
@@ -283,32 +285,58 @@ public boolean isErrorDetected() {
...
      */
-    protected void computeSamplerName(HTTPSamplerBase sampler,
-            HttpRequestHdr request) {
-        String prefix = request.getPrefix();
+    protected void computeSamplerName(HTTPSamplerBase sampler, HttpRequestHdr request) {
+        String prefix = request.getPrefix(); // ppp
         int httpSampleNameMode = request.getHttpSampleNameMode();
         if (!HTTPConstants.CONNECT.equals(request.getMethod()) && isNumberRequests()) {
             if(StringUtils.isNotEmpty(prefix)) {
+            	// with a prefix name
```
Comment (FSchumacher) [inline src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:293]: Can these parts be extracted in smaller functions/methods? You could use names to tell the reader what the function is doing instead of using a comment and at the same time the code gets less (at least here :) )

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java [hunk 7]
```diff
@@ -283,32 +285,58 @@ public boolean isErrorDetected() {
      * @param sampler {@link HTTPSamplerBase}
      * @param request {@link HttpRequestHdr}
      */
-    protected void computeSamplerName(HTTPSamplerBase sampler,
-            HttpRequestHdr request) {
-        String prefix = request.getPrefix();
+    protected void computeSamplerName(HTTPSamplerBase sampler, HttpRequestHdr request) {
+        String prefix = request.getPrefix(); // ppp
```
Comment (FSchumacher) [inline src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:289]: What is the meaning of *ppp* ?

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/AbstractSamplerCreator.java [hunk 8]
```diff
@@ -86,30 +75,49 @@ public AbstractSamplerCreator() {
     /**
      * @return int request number
      */
-    protected static int getRequestNumber() {
+    public int getRequestNumber() {
```
Comment (FSchumacher) [inline src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/AbstractSamplerCreator.java:78]: Why did you change it to be non-static? (I am not sure, I like the static, but I would like to here your reasoning.

File: bin/jmeter.properties [hunk 9]
```diff
@@ -604,8 +604,15 @@ upgrade_properties=/bin/upgrade.properties
 # it assumes that the user has clicked a new URL
 #proxy.pause=5000
 
-# Add numeric suffix to Sampler names (default true)
+# Add numeric to Sampler names (default true)
```
Comment (FSchumacher) [inline bin/jmeter.properties:607]: Some placeholder has to be used (in my opinion) instead of *suffix* or use *number* instead of *numeric*.

File: xdocs/usermanual/component_reference.xml (hunk 3 above)
Comment (vdaburon) [reply xdocs/usermanual/component_reference.xml:6722]: I don't change this line, i get it from a previous version before the modiifcation, but you are rigth, the correct <code>proxy.headers.remove</code>

File: xdocs/usermanual/component_reference.xml (hunk 5 above)
Comment (vdaburon) [reply xdocs/usermanual/component_reference.xml:6730]: ok i will use e.g.

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java (hunk 6 above)
Comment (vdaburon) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:293]: I add a comment because "prefix" have 2 significations :
 - prefix name
 - prefix numbering 
This is prefix name.

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java (hunk 7 above)
Comment (vdaburon) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/DefaultSamplerCreator.java:289]: ppp is "ppp"refix name" i use this string to test (look at the image for the description of the modification and the test.
nnn is "nnn"ame i use this string to test the naming request.

File: bin/jmeter.properties (hunk 9 above)
Comment (vdaburon) [reply bin/jmeter.properties:607]: I remove the suffix (add to the JMeter 5.2) because is not only suffix.
I don't change the comment "numeric". This comment exist for a long time. 
I can change to numbering.

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/gui/ProxyControlGui.java (hunk 2 above)
Comment (vdaburon) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/gui/ProxyControlGui.java:487]: I use prefix  i for Integer or s for String for local variables but i could change to requestNumber

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/AbstractSamplerCreator.java (hunk 8 above)
Comment (vdaburon) [reply src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/AbstractSamplerCreator.java:78]: The method need to be public not protected (same package)
This method return value from a static value but i don't see why this method need to be static.
Other methods are not static.

**Comment by FSchumacher** [comment]
Have you had a chance to look at #595?
Do you really switch between naming schemes while recording, or is this just to show the possibilities? 
Should the setting of _next number_  be more like a command? Like having a button next to it saying: 'reset counter to..'?

File: src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/ProxyControl.java [hunk 10]
```diff
@@ -355,6 +406,48 @@ public void setHTTPSampleNamingMode(int httpNamingMode) {
...
+    public void setHttpSamplerNumberingStartForSamplerCreator(int httpSamplerNumberingStartValue) {
+        iNumberingStartValue =  httpSamplerNumberingStartValue;
+        // change the numbering start value directly to the AtomicInteger REQUEST_NUMBER in the AbstractSamplerCreator
+        SamplerCreator samplerCreator = samplerCreatorFactory.getDefaultSamplerCreator();
+        samplerCreator.setRequestNumber(httpSamplerNumberingStartValue);
+    }
+
+    public int getRequestNumberFromSamplerCreator() {
+        int iRequestNumber = 1;
+        SamplerCreator samplerCreator = samplerCreatorFactory.getDefaultSamplerCreator();
```
Comment (FSchumacher) [inline src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/proxy/ProxyControl.java:440]: I think that this assumption may not always be correct, as the `DefaultSamplerCreator` is probably the only factory, that would really work here.
On the other hand that seems to be the same with the formatting scheme changes. Hm...

**Comment by FSchumacher** [comment]
Thanks for your contribution on this feature. I hope you like the implemented feature, even if not all of your code has made it into the codebase. It would be great, if you could check, whether a current nightly works for you.
